### PR TITLE
KAZOO-4314 freeswitch channels API

### PR DESCRIPTION
**Pull request by lazedo**

### CODE REVIEW

File: applications/ecallmgr/src/ecallmgr_fs_channel.erl [hunk 1]
```diff
@@ -387,23 +431,69 @@ code_change(_OldVsn, State, _Extra) ->
...
+               ,{<<"sip_h_X-ecallmgr_Account-ID">>, props:get_value(<<"Account-ID">>, ChannelVars)}
+               ,{<<"sip_h_X-ecallmgr_Realm">>, props:get_value(<<"Realm">>, ChannelVars)}
+              ]),
+    X = string:join(
+          [wh_util:to_list(<<K/binary, "='", (wh_util:to_binary(V))/binary, "'">>)
+           || {K, V}  <- Props
+          ]
+          ,","
+         ),
+    wh_util:to_binary(["[", X, "]"]).
```
Comment (fenollp) [inline applications/ecallmgr/src/ecallmgr_fs_channel.erl]: I'm confused by all these conversionnnnns
```erlang
fs_props_to_string([]) -> <<"[]">>; %% dunno if useful
fs_props_to_string([{Hk,Hv}|T]) ->
    Rest = << <<",", K/binary, "='", (wh_util:to_list(V))/binary, "'">> || {K,V} <- T >>,
    <<"[", Hk/binary, "='", (wh_util:to_binary(Hv))/binary, "'", Rest/binary, "]">>.
```

…much more readable, isn't it? :d

File: core/whistle-1.0.0/src/wh_util.erl [hunk 2]
```diff
@@ -632,6 +633,12 @@ spawn(Module, Function, Arguments) ->
                          _ = put_callid(CallId),
                          erlang:apply(Module, Function, Arguments)
                  end).
+spawn(Fun, Arguments) ->
+    CallId = get_callid(),
+    erlang:spawn(fun() ->
+                         _ = put_callid(CallId),
+                         erlang:apply(Fun, Arguments)
+                 end).
```
Comment (fenollp) [inline core/whistle-1.0.0/src/wh_util.erl]: I really like this!

Remove `spawn/3` and replace each of its calls with `spawn/2` and we'll have full coverage with xref!

File: applications/ecallmgr/src/ecallmgr_fs_channel.erl [hunk 3]
```diff
@@ -258,6 +258,46 @@ to_api_props(Channel) ->
        ,{<<"Elapsed-Seconds">>, wh_util:elapsed_s(Channel#channel.timestamp)}
       ]).
 
+-spec channel_ccvs(channel() | wh_json:object() | wh_proplist()) -> wh_proplist().
+channel_ccvs(Channel)
```
Comment (jamesaimonetti) [inline applications/ecallmgr/src/ecallmgr_fs_channel.erl]: You can match the record in the function header:

```erlang
channel_ccvs(#channel{}=Channel) -> ...;
channel_ccvs([_|_]=Props) -> ...;
channel_ccvs(JObj) -> ...
```

File: applications/ecallmgr/src/ecallmgr_fs_channel.erl [hunk 4]
```diff
@@ -387,23 +431,68 @@ code_change(_OldVsn, State, _Extra) ->
...
+               ,{<<"sip_h_X-ecallmgr_refer-from-channel-id">>, props:get_value(<<"refer-from-channel-id">>, ReqProps)}
+               ,{<<"sip_h_X-ecallmgr_refer-for-channel-id">>, props:get_value(<<"refer-for-channel-id">>, ReqProps)}
+               ,{<<"sip_h_X-ecallmgr_Account-ID">>, props:get_value(<<"Account-ID">>, ChannelVars)}
+               ,{<<"sip_h_X-ecallmgr_Realm">>, props:get_value(<<"Realm">>, ChannelVars)}
+              ]),
+    fs_props_to_binary(Props).
+
+-spec fs_props_to_binary(wh_proplist()) -> ne_binary().
+fs_props_to_binary([{Hk,Hv}|T]) ->
+    Rest = << <<",", K/binary, "='", wh_util:to_list(V), "'">> || {K,V} <- T >>,
```
Comment (jamesaimonetti) [inline applications/ecallmgr/src/ecallmgr_fs_channel.erl]: Change `wh_util:to_list(V)` to `(wh_util:to_binary(V))/binary`; same on the next line.

File: core/whistle-1.0.0/src/wh_util.erl (hunk 2 above)
Comment (lazedo) [inline core/whistle-1.0.0/src/wh_util.erl]: please create a ticket for this, "replace usage of wh_util spawn/3 with spawn/2 on all applications"

**Comment by jamesaimonetti** [comment]
@lazedo was there anything left to do here? LGTM. @k-anderson all clear from you?

**Comment by k-anderson** [comment]
@jamesaimonetti agreed! I was just waiting to see if there was anything left from @fenollp comment ;)

File: core/whistle-1.0.0/src/wh_util.erl (hunk 2 above)
Comment (fenollp) [inline core/whistle-1.0.0/src/wh_util.erl]: Done!

**Pull request merged**